File: c1_algorithmic_toolbox/w5_dynamic_programming/edit_distance.py
# Uses python3
import random


def edit_distance(A, B):
    n = len(A)
    m = len(B)
    D = [[0 for j in range(m + 1)] for i in range(n + 1)]
    for i in range(n+1):
        D[i][0] = i
    for j in range(m+1):
        D[0][j] = j

    for j in range(1, m+1):
        for i in range(1, n+1):
            insertion = D[i][j-1] + 1
            deletion = D[i-1][j] + 1
            match = D[i-1][j-1]
            mismatch = D[i-1][j-1] + 1
            if A[i-1] == B[j-1]:
                D[i][j] = min(insertion, deletion, match)
            else:
                D[i][j] = min(insertion, deletion, mismatch)

    # # Print Distance Matrix D
    # for i in range(n+1):
    #     for j in range(m+1):
    #         print('{0:4}'.format(D[i][j]), end=' ')
    #     print()

    return D[n][m]

if __name__ == "__main__":
    print(edit_distance(input(), input()))

    # # Time test
    # alphabet = u'0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    # A = [random.choice(alphabet) for i in range(100)]
    # B = [random.choice(alphabet) for i in range(100)]
    # print(edit_distance(A, B))

File: c1_algorithmic_toolbox/w5_dynamic_programming/edit_distance.cpp
#include <iostream>

#include <string>

 

using namespace std;

 

int

edit_distance (const string & str1, const string & str2)

{

  int n = str1.size ();

  int m = str2.size ();

  int D[n + 1][m + 1];

  int i = 0;

 

  while (i <= n)

    {

      D[i][0] = i;

      i++;

    }

 

  i = 1;

  while (i <= m)

    {

      D[0][i] = i;

      i++;

    }

 

  for (int i = 1; i <= n; i++)

    {

      for (int j = 1; j <= m; j++)

                {

                  int insertion = D[i][j - 1] + 1;

                  int deletion = D[i - 1][j] + 1;

                  int match = D[i - 1][j - 1];

                  int mismatch = D[i - 1][j - 1] + 1;

                  if (str1[i - 1] == str2[j - 1])

                    {

                      D[i][j] = min (insertion, min (deletion, match));

                    }

                  else

                    {

                      D[i][j] = min (insertion, min (deletion, mismatch));

                    }

 

                }

    }

 

  return D[n][m];

}

 

int

main ()

{

  string str1;

  string str2;

  cin >> str1 >> str2;

  cout << edit_distance (str1, str2) << endl;

  return 0;

}